### PR TITLE
drivers: nrf_wifi: Restructure to carve out a new nrf70 bus lib

### PR DESCRIPTION
**Pull request by bama-nordic**
The nrf wifi drivers include code related to qspi/spi bus. This work moves bus related files to a new nrf70 bus library which is now independent of the core driver and facilitlates building applications directly using the bus APIs, agnostic to enabling wifi configs.

### CODE REVIEW

File: modules/nrf_wifi/bus/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+  zephyr_library_sources(rpu_hw_if.c)
+
+  zephyr_library_sources_ifdef(CONFIG_NRF70_ON_QSPI
+    qspi_if.c
+  )
+  zephyr_library_sources_ifdef(CONFIG_NRF70_ON_SPI
+    spi_if.c
+  )
+  zephyr_library_sources(device.c)
```
Comment (krish2718) [inline modules/nrf_wifi/bus/CMakeLists.txt:28]: this can be combined with `rpu_hw_if.c`

File: modules/nrf_wifi/bus/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+# Check if ZEPHYR_BASE is set
+if(NOT DEFINED ENV{ZEPHYR_BASE})
+    message(FATAL_ERROR "ZEPHYR_BASE environment variable is not set. Please set it to the Zephyr base directory.")
+endif()
+
+if (CONFIG_NRF70_BUSLIB)
+
+  zephyr_library_named(nrf70-buslib)
+
```
Comment (krish2718) [inline modules/nrf_wifi/bus/CMakeLists.txt:15]: no need for empty lines

File: modules/nrf_wifi/bus/Kconfig [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+# Hidden symbols for internal use
+config WIFI_NRF7002
+        bool
+        default y if DT_HAS_NORDIC_NRF7002_SPI_ENABLED || DT_HAS_NORDIC_NRF7002_QSPI_ENABLED
+
+config WIFI_NRF7001
+        bool
+        default y if DT_HAS_NORDIC_NRF7001_SPI_ENABLED || DT_HAS_NORDIC_NRF7001_QSPI_ENABLED
+
+config WIFI_NRF7000
```
Comment (krish2718) [inline modules/nrf_wifi/bus/Kconfig:27]: these are only used in native driver

File: modules/nrf_wifi/bus/Kconfig [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+	def_bool $(dt_compat_enabled,$(DT_COMPAT_NORDIC_NRF70_QSPI))
+	select NRFX_QSPI
+
+config NRF70_ON_SPI
+	def_bool $(dt_compat_enabled,$(DT_COMPAT_NORDIC_NRF70_SPI))
+	select SPI
+
+#config NRF70_QSPI_LOW_POWER
+#	bool "low power mode in QSPI"
+#	default y if NRF_WIFI_LOW_POWER
```
Comment (krish2718) [inline modules/nrf_wifi/bus/Kconfig:42]: local change?

File: modules/nrf_wifi/bus/rpu_hw_if.c [hunk 5]
```diff
@@ -483,10 +483,12 @@ int rpu_enable(void)
 		goto rpu_pwroff;
 	}
 
+#if 0
```
Comment (krish2718) [inline modules/nrf_wifi/bus/rpu_hw_if.c:486]: local change?

File: modules/nrf_wifi/bus/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  zephyr_library_sources_ifdef(CONFIG_NRF70_ON_QSPI
+    qspi_if.c
+  )
+  zephyr_library_sources_ifdef(CONFIG_NRF70_ON_SPI
+    spi_if.c
+  )
+  zephyr_library_sources(device.c)
+
+  set_target_properties(nrf70-buslib PROPERTIES INTERFACE_INCLUDE_DIRECTORIES ${CMAKE_CURRENT_SOURCE_DIR})
```
Comment (krish2718) [inline modules/nrf_wifi/bus/CMakeLists.txt:30]: `zephyr_library_include_directories` should take care of this automatically.

**Review by krish2718 — COMMENTED** [review]
remove `ficr_prog.*` should be a separate commit as it's unrelated to buslib.

File: modules/nrf_wifi/bus/Kconfig (hunk 4 above)
Comment (bama-nordic) [reply modules/nrf_wifi/bus/Kconfig:42]: moved to NRF70 driver Kconfig

File: modules/nrf_wifi/bus/rpu_hw_if.c (hunk 5 above)
Comment (bama-nordic) [reply modules/nrf_wifi/bus/rpu_hw_if.c]: We cannot have `rpu_validate_comms()` anymore since it will not work unless firmware is loaded and for a independent bus lib, this wont work. I think its best we move this function as we will have bustest now, plan to do in the next commit

Comment (bama-nordic) [reply modules/nrf_wifi/bus/rpu_hw_if.c]: OK, I have put this under #ifdef CONFIG_WIFI_NRF70 - that should take care of the issue while not changing functionality for the core wifi driver.

File: modules/nrf_wifi/bus/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,28 @@
+#
+# Copyright (c) 2024 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: BSD-3-Clause
```
Comment (krish2718) [inline modules/nrf_wifi/bus/CMakeLists.txt:4]: this should also be apache-2.0 for scan code to pass.

File: modules/nrf_wifi/bus/Kconfig [hunk 8]
```diff
@@ -0,0 +1,38 @@
+#
+# Copyright (c) 2024 Nordic Semiconductor
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+DT_COMPAT_NORDIC_NRF70_QSPI := nordic,nrf7002-qspi
```
Comment (krish2718) [inline modules/nrf_wifi/bus/Kconfig:7]: This won't work with nRF7000/nRF7001, need to add entries for both.

Comment (bama-nordic) [reply modules/nrf_wifi/bus/Kconfig]: Added both now

**Review by krish2718 — DISMISSED** [review]

**Review by krish2718 — DISMISSED** [review]

File: modules/nrf_wifi/bus/qspi_if.c [hunk 9]
```diff
@@ -195,7 +190,7 @@ struct qspi_nor_data {
 	struct k_sem sync;
 	/* The semaphore to control driver init/uninit. */
 	struct k_sem count;
-#else /* CONFIG_MULTITHREADING */
+#else  /* CONFIG_MULTITHREADING */
```
Comment (jukkar) [inline modules/nrf_wifi/bus/qspi_if.c:193]: Could you remove these unnecessary changes. I see it here and below in multiple places.

File: modules/nrf_wifi/bus/qspi_if.c [hunk 10]
```diff
@@ -688,7 +681,7 @@ static inline void qspi_fill_init_struct(nrfx_qspi_config_t *initstruct)
 #endif
 #endif /* CONFIG_PINCTRL */
 	/* Configure Protocol interface */
-	initstruct->prot_if.addrmode = NRF_QSPI_ADDRMODE_24BIT;
+		initstruct->prot_if.addrmode = NRF_QSPI_ADDRMODE_24BIT;
```
Comment (jukkar) [inline modules/nrf_wifi/bus/qspi_if.c:684]: Indent is now wrong

File: modules/nrf_wifi/bus/qspi_if.c [hunk 11]
```diff
@@ -1000,28 +993,23 @@ static int qspi_nor_configure(const struct device *dev)
 static int qspi_nor_init(const struct device *dev)
 {
 #ifdef CONFIG_PINCTRL
-	int ret = pinctrl_apply_state(QSPI_IF_BUS_PINCTRL_DT_DEV_CONFIG_GET,
-				      PINCTRL_STATE_DEFAULT);
+	int ret = pinctrl_apply_state(QSPI_IF_BUS_PINCTRL_DT_DEV_CONFIG_GET, PINCTRL_STATE_DEFAULT);
```
Comment (jukkar) [inline modules/nrf_wifi/bus/qspi_if.c:996]: Did you run this through clang-format? It does these unnecessary changes, please remove them.

**Review by jukkar — CHANGES_REQUESTED** [review]
Could you remove all cosmetic and unnecessary changes, they make it hard to review this.

**Review by krish2718 — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

File: modules/nrf_wifi/bus/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
```
Comment (fabiobaltieri) [inline modules/nrf_wifi/bus/CMakeLists.txt:4]: Hi, this has not been fixed and is still BSD-3-Clause upstream, @krish2718 would you be able to send a PR to fix it?

https://github.com/zephyrproject-rtos/zephyr/blob/c7a19da8ca1abefa8eac41e704e7c678af0282a9/modules/nrf_wifi/bus/CMakeLists.txt#L4

cc @kartben

Comment (krish2718) [reply modules/nrf_wifi/bus/CMakeLists.txt:4]: FYI https://github.com/zephyrproject-rtos/zephyr/pull/86748
